File: backend/services/encounter-service/app/models/encounter.py
from typing import Dict, List, Optional, Any
from pydantic import BaseModel, Field
from datetime import datetime
from enum import Enum

# Import shared FHIR models
from shared.models import (
    Encounter, CodeableConcept, Reference, Period,
    EncounterParticipant, EncounterDiagnosis, EncounterLocation
)

class EncounterStatus(str, Enum):
    """Encounter status based on FHIR standard"""
    PLANNED = "planned"
    ARRIVED = "arrived"
    TRIAGED = "triaged"
    IN_PROGRESS = "in-progress"
    ONLEAVE = "onleave"
    FINISHED = "finished"
    CANCELLED = "cancelled"
    ENTERED_IN_ERROR = "entered-in-error"
    UNKNOWN = "unknown"

class EncounterClass(str, Enum):
    """Encounter class based on FHIR standard"""
    AMBULATORY = "AMB"
    EMERGENCY = "EMER"
    INPATIENT = "IMP"
    OUTPATIENT = "OUTPATIENT"
    VIRTUAL = "VR"
    HOME_HEALTH = "HH"

class EncounterCreate(BaseModel):
    """Model for creating an encounter"""
    status: EncounterStatus
    class_value: Dict[str, str] = Field(..., alias="class")
    type: Optional[List[CodeableConcept]] = None
    subject: Reference
    participant: Optional[List[EncounterParticipant]] = None
    period: Optional[Period] = None
    reasonCode: Optional[List[CodeableConcept]] = None
    diagnosis: Optional[List[EncounterDiagnosis]] = None
    location: Optional[List[EncounterLocation]] = None
    serviceProvider: Optional[Reference] = None

    class Config:
        schema_extra = {
            "example": {
                "status": "in-progress",
                "class": {
                    "system": "http://terminology.hl7.org/CodeSystem/v3-ActCode",
                    "code": "AMB",
                    "display": "ambulatory"
                },
                "type": [
                    {
                        "coding": [
                            {
                                "system": "http://snomed.info/sct",
                                "code": "308335008",
                                "display": "Patient encounter procedure"
                            }
                        ],
                        "text": "Outpatient visit"
                    }
                ],
                "subject": {
                    "reference": "Patient/123",
                    "display": "John Smith"
                },
                "period": {
                    "start": "2023-06-15T08:00:00Z"
                }
            }
        }

    def to_fhir_encounter(self) -> Encounter:
        """Convert to a FHIR Encounter."""
        data = self.model_dump(exclude_unset=True)

        # Convert status enum to string
        if isinstance(data.get('status'), EncounterStatus):
            data['status'] = data['status'].value

        # Handle class field
        if 'class_value' in data:
            data['class'] = data.pop('class_value')

        return Encounter(**data)

class EncounterUpdate(BaseModel):
    """Model for updating an encounter"""
    status: Optional[EncounterStatus] = None
    class_value: Optional[Dict[str, str]] = Field(None, alias="class")
    type: Optional[List[CodeableConcept]] = None
    subject: Optional[Reference] = None
    participant: Optional[List[EncounterParticipant]] = None
    period: Optional[Period] = None
    reasonCode: Optional[List[CodeableConcept]] = None
    diagnosis: Optional[List[EncounterDiagnosis]] = None
    location: Optional[List[EncounterLocation]] = None
    serviceProvider: Optional[Reference] = None

    def to_fhir_encounter_update(self) -> Dict[str, Any]:
        """Convert to a FHIR Encounter update."""
        data = self.model_dump(exclude_unset=True)

        # Convert status enum to string
        if isinstance(data.get('status'), EncounterStatus):
            data['status'] = data['status'].value

        # Handle class field
        if 'class_value' in data:
            data['class'] = data.pop('class_value')

        return data

class EncounterInDB(BaseModel):
    """Model for an encounter in the database"""
    id: str
    resourceType: str = "Encounter"
    status: EncounterStatus
    class_value: Dict[str, str] = Field(..., alias="class")
    type: Optional[List[CodeableConcept]] = None
    subject: Reference
    participant: Optional[List[EncounterParticipant]] = None
    period: Optional[Period] = None
    reasonCode: Optional[List[CodeableConcept]] = None
    diagnosis: Optional[List[EncounterDiagnosis]] = None
    location: Optional[List[EncounterLocation]] = None
    serviceProvider: Optional[Reference] = None

    @classmethod
    def from_fhir_encounter(cls, encounter: Encounter) -> 'EncounterInDB':
        """Create from a FHIR Encounter."""
        data = encounter.model_dump(exclude_unset=True)

        # Handle class field
        if 'class' in data:
            data['class_value'] = data.pop('class')

        return cls(**data)

File: vaidshala/clinical-runtime-platform/services/intake-onboarding-service/internal/fhir/encounter.go
package fhir

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// NewEncounterResource creates a FHIR R4 Encounter resource for an intake session.
func NewEncounterResource(patientID uuid.UUID, encounterType string) ([]byte, error) {
	encounter := map[string]interface{}{
		"resourceType": "Encounter",
		"status":       "in-progress",
		"class": map[string]interface{}{
			"system":  "http://terminology.hl7.org/CodeSystem/v3-ActCode",
			"code":    "VR",
			"display": "virtual",
		},
		"type": []map[string]interface{}{
			{
				"coding": []map[string]interface{}{
					{
						"system":  "http://cardiofit.in/encounter-types",
						"code":    encounterType,
						"display": encounterType + " session",
					},
				},
			},
		},
		"subject": map[string]interface{}{
			"reference": fmt.Sprintf("Patient/%s", patientID),
		},
		"period": map[string]interface{}{
			"start": time.Now().UTC().Format(time.RFC3339),
		},
	}

	return json.Marshal(encounter)
}

// UpdateEncounterStatus updates the status of an existing Encounter.
func UpdateEncounterStatus(existingEncounter []byte, status string) ([]byte, error) {
	var encounter map[string]interface{}
	if err := json.Unmarshal(existingEncounter, &encounter); err != nil {
		return nil, err
	}

	encounter["status"] = status
	if status == "finished" {
		if period, ok := encounter["period"].(map[string]interface{}); ok {
			period["end"] = time.Now().UTC().Format(time.RFC3339)
		}
	}

	return json.Marshal(encounter)
}
